Firestore/core/test/unit/model/values_test.cc

```diff
@@ -0,0 +1,323 @@
+                      ComparisonResult cmp) {
+    for (const auto& val1 : left) {
+      for (const auto& val2 : right) {
+        EXPECT_EQ(cmp, Values::Compare(val1, val2));
+      }
+    }
+  }
+
+  void VerifyCanonicalId(const google_firestore_v1_Value& value,
+                         const std::string& expected_canonical_id) {
+    const std::string& actual_canonical_id = Values::CanonicalId(value);
+    EXPECT_EQ(expected_canonical_id, actual_canonical_id);
+  }
+
+ private:
+  remote::Serializer serializer;
+};
+
+TEST_F(ValuesTest, Equality) {
+  std::vector<std::vector<google_firestore_v1_Value>> equals_group;
+
+  Add(equals_group, Wrap(nullptr), Wrap(nullptr));
+  Add(equals_group, Wrap(false), Wrap(false));
+  Add(equals_group, Wrap(true), Wrap(true));
+  Add(equals_group, Wrap(std::numeric_limits<double>::quiet_NaN()),
+      Wrap(ToDouble(kCanonicalNanBits)), Wrap(ToDouble(kNanBits)),
+      Wrap(std::nan("1")), Wrap(std::nan("2")));
+  // -0.0 and 0.0 compare the same but are not equal.
+  Add(equals_group, Wrap(-0.0));
+  Add(equals_group, Wrap(0.0));
+  Add(equals_group, Wrap(1), Wrap(1LL));
+  // Doubles and Longs aren't equal (even though they compare same).
+  Add(equals_group, Wrap(1.0), Wrap(1.0));
+  Add(equals_group, Wrap(1.1), Wrap(1.1));
+  Add(equals_group, Wrap(BlobValue(0, 1, 1)));
+  Add(equals_group, Wrap(BlobValue(0, 1)));
+  Add(equals_group, Wrap("string"), Wrap("string"));
+  Add(equals_group, Wrap("strin"));
+  // latin small letter e + combining acute accent
+  Add(equals_group, Wrap("e\u0301b"));
+  // latin small letter e with acute accent
+  Add(equals_group, Wrap("\u00e9a"));
+  Add(equals_group, Wrap(Timestamp::FromTimePoint(kDate1)), Wrap(kTimestamp1));
+  Add(equals_group, Wrap(Timestamp::FromTimePoint(kDate2)), Wrap(kTimestamp2));
+  // NOTE: ServerTimestampValues can't be parsed via Wrap().
+  Add(equals_group,
+      WrapServerTimestamp(FieldValue::FromServerTimestamp(kTimestamp1)),
+      WrapServerTimestamp(FieldValue::FromServerTimestamp(kTimestamp1)));
+  Add(equals_group,
+      WrapServerTimestamp(FieldValue::FromServerTimestamp(kTimestamp2)));
+  Add(equals_group, Wrap(GeoPoint(0, 1)), Wrap(GeoPoint(0, 1)));
+  Add(equals_group, Wrap(GeoPoint(1, 0)));
+  Add(equals_group, WrapReference(DbId(), Key("coll/doc1")),
+      WrapReference(DbId(), Key("coll/doc1")));
+  Add(equals_group, WrapReference(DbId(), Key("coll/doc2")));
+  Add(equals_group, WrapReference(DbId("project/baz"), Key("coll/doc2")));
+  Add(equals_group, WrapArray("foo", "bar"), WrapArray("foo", "bar"));
+  Add(equals_group, WrapArray("foo", "bar", "baz"));
+  Add(equals_group, WrapArray("foo"));
+  Add(equals_group, WrapObject("bar", 1, "foo", 2),
+      WrapObject("foo", 2, "bar", 1));
+  Add(equals_group, WrapObject("bar", 2, "foo", 1));
+  Add(equals_group, WrapObject("bar", 1));
+  Add(equals_group, WrapObject("foo", 1));
+
+  for (size_t i = 0; i < equals_group.size(); ++i) {
```

Comment:
It took me a while to understand why the tests are written like this..maybe some comment will help to explain what is happening, something like:

Each element of `equals_group` is a vector of values are should equal to each other. Any two values that do no belong in the same `equals_group` element should not equal to each other.

Reply:
Also consider break the equal/NotEqual check in two separate loop, more readable IMO.

Reply:
I added comments and rewrote the loop a bit. Do you think it is more understandable now?

Reply:
Yep!